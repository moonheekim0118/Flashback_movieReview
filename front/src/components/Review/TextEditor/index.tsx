import React , { 
       useCallback , 
       useRef, 
       useState , 
       useEffect 
} from 'react';
import useToggle from '../../../hooks/useSelecting';
import Router from 'next/router';
import { useDispatch , useSelector } from 'react-redux';
import MovieCard from '../../Movie/MovieCard';
import Button from '../../../atoms/Buttons';
import Icon from '../../../atoms/Icons';
import styled from 'styled-components';
import useValidation from '../../../hooks/useValidation';
import Badge from '../Badge';
import { faCheck , faTimes } from '@fortawesome/free-solid-svg-icons';
import { ADD_MY_REVIEW_REQUEST, UPDATE_MY_REVIEW_REQUEST,} from '../../../actions/review';
import { OPEN_ALERT } from '../../../actions/alert';
import { ReviewList } from '../../../model/ReviewList';

interface Props {
    Review?:ReviewList;
    ButtonType:string;
}

const TextEditor=({Review , ButtonType}:Props)=>{
    const dispatch = useDispatch();

    // 아이콘 
    const PassedIcon=<Icon icon={faCheck} color={'green'}/>
    const ErrorIcon=<Icon icon={faTimes} color={'red'}/>

    const [ shortComment,  // 짧은 평 
            setShortComment, 
            shortCommentError 
        ] = useValidation(Review.shortComment,5,20);

    const [ character,  // 기억에 남는 캐릭터 
            setCharacter, 
            characterError 
        ] = useValidation(Review.character,5,50);

    const [ line,  // 기억에 남는 대사 
            setLine, 
            lineError 
        ] = useValidation(Review.line,5,50);

    const [ scene,  // 기억에 남는 장면 
            setScene, 
            sceneError 
        ] = useValidation(Review.scene,5,50);
        
    const [ freeComment, // 자유 커멘트 
            setFreeComment,
            freeCommentError 
        ] = useValidation(Review.freeComment,0,50);

    const initialUpdate = useRef(true);// 최초 렌더링인지 구분하기 위함 
    const [initial, setInitial]=useState(true); // 최초 렌더링인지 구분하기 위함 
    
    // 수정 상태 시 원래 저장된 레이팅으로 상태 설정
    const good = Review.rating==='GOOD' ? true : false;
    const soso = Review.rating==='SOSO' ? true : false;
    const bad = Review.rating==='BAD' ? true : false; 

    const [goodSelect,setGoodSelect,
           sosoSelect,setSoSoSelect,
           badSelect,setBadSelect]
         = useToggle(good, soso, bad);

    const { myReviews ,
            addMyReviewDone, 
            addMyReviewError, 
            updateMyReviewDone, 
            updateMyReviewError } 
            = useSelector((state)=>state.review);

    useEffect(()=>{ // 초기 상태라면 validation 에러 띄워주지 않음 
        if(initialUpdate.current){
            initialUpdate.current=false;
            return;
        }
        else{
            setInitial(false);
        }
    },[initialUpdate.current]);

    useEffect(()=>{ // 리뷰 생성 후 alert 보여준 후 리다이렉트 
        if(addMyReviewDone){
            dispatch({type:OPEN_ALERT,data:"리뷰가 등록되었습니다."}); // open alert 
            const  timer = setTimeout(()=> Router.replace(`/singleReview/${myReviews[0].id}`),5000); // 리다이렉트
            return ()=>clearTimeout(timer);
        }
        else if(addMyReviewError){ // 에러 발생 시 에러 메시지 띄워줌 
            dispatch({type:OPEN_ALERT,data:addMyReviewError});
        }
    },[addMyReviewDone,addMyReviewError]);

    useEffect(()=>{ // 리뷰 수정 후 alert 보여준 후 리다이렉트
        if(updateMyReviewDone){
            dispatch({type:OPEN_ALERT,data:"리뷰가 수정되었습니다."});
            const timer = setTimeout(()=>Router.replace(`/singleReview/${Review.id}`),5000);
            return ()=>clearTimeout(timer);
        }
        else if(updateMyReviewError){ // 에러 발생시 에러 메시지 띄워줌 
            dispatch({type:OPEN_ALERT,data:updateMyReviewError});
        }
    },[updateMyReviewDone,updateMyReviewError]);

    const onCreate=useCallback(()=>{ // 리뷰 저장
        let rating='BAD'; // 레이팅 설정 
        if(goodSelect){
            rating='GOOD';
        } else if(sosoSelect){
            rating='SOSO'
        }
        dispatch({
            type:ADD_MY_REVIEW_REQUEST,
            data:{
                Movie:Review.Movie,
                rating:rating,
                shortComment:shortComment,
                character:character,
                line:line,
                scene:scene,
                freeComment:freeComment,
            }
        });
    },[goodSelect,sosoSelect,badSelect,shortComment,character,line,scene,freeComment]);

    const onUpdate=useCallback(()=>{ // 리뷰 수정 
        let rating='BAD';
        if(goodSelect){
            rating='GOOD';
        } else if(sosoSelect){
            rating='SOSO'
        }
        dispatch({
            type:UPDATE_MY_REVIEW_REQUEST,
            data:{
                id:Review.id,
                User:Review.User,
                Movie:Review.Movie,
                rating:rating,
                shortComment:shortComment,
                line:line,
                character:character,
                scene:scene,
                freeComment:freeComment,
            }
        })
    },[goodSelect,sosoSelect,badSelect,shortComment,character,line,scene,freeComment]);

    // 버튼 disabeld 공통 조건 
    const disabledRequirements= 
    shortCommentError||characterError||lineError||sceneError||freeCommentError||shortComment.length===0||
    character.length===0||line.length===0|| scene.length===0; 

    // 업데이트 버튼 disabeld 조건 
    const updateDisabledRequirements=
    shortComment===Review.shortComment && freeComment===Review.freeComment && 
    character===Review.character && line===Review.line && scene===Review.scene ; 
    const SubmitButton= ButtonType==='create' ?
    <Button
    color={"purple"}
    onClick={onCreate}
    title={"저장하기"}
    disabled={disabledRequirements }
    /> : 
    <Button
    color={"purple"}
    onClick={onUpdate}
    title={"수정하기"}
    disabled={disabledRequirements || updateDisabledRequirements}
    /> ;

    return(
        <Container>
            <MovieCard Movie={Review.Movie}/>
            <BadgeContainer>
                <Badge badgeName={"GOOD"} selected={goodSelect} onClick={setGoodSelect}/>
                <Badge badgeName={"SOSO"} selected={sosoSelect} onClick={setSoSoSelect}/>
                <Badge badgeName={"BAD"} selected={badSelect} onClick={setBadSelect} />
            </BadgeContainer>
            <TextContainer>
                <Question htmlFor="shortComment">
                    한 줄평
                </Question>
                {shortCommentError && ErrorIcon}
                {!shortCommentError && !initial && shortComment.length >0 && PassedIcon}
                <TextInput 
                value={shortComment}
                onChange={setShortComment}
                name="shortComment"
                />
                <Question htmlFor="character"> 
                    기억에 남는 인물과 이유를 적어주세요
                </Question>
                {characterError && ErrorIcon}
                {!characterError && !initial && character.length>0 && PassedIcon}
                <TextInput
                value={character}
                onChange={setCharacter}
                name="character"
                />
                <Question htmlFor="line">
                    기억에 남는 대사와 이유를 적어주세요
                </Question>
                {lineError && ErrorIcon}
                {!lineError && !initial && line.length > 0 && PassedIcon}
                <TextInput
                value={line}
                onChange={setLine}
                name="line"
                />
                <Question htmlFor="scene">
                    기억에 남는 장면과 이유를 적어주세요
                </Question>
                {sceneError && ErrorIcon}
                {!sceneError && !initial && scene.length > 0 && PassedIcon}
                <TextInput
                value={scene}
                onChange={setScene}
                name="scene"
                />
                <Question htmlFor="freeComment">
                    영화에 대한 자유로운 메모 
                </Question>
                {freeCommentError && ErrorIcon}
                {!freeCommentError && !initial && freeComment.length>0 && PassedIcon}
                <TextInput
                value={freeComment}
                onChange={setFreeComment}
                name="freeComment"
                />
            </TextContainer>
            <ButtonWrapper>
                {SubmitButton}
            </ButtonWrapper>
        </Container>
    );
}

const Container = styled.div`
    width:80%;
    margin:auto;
    background-color:inherit;
`;
const Question = styled.label`
    font-size:1.5rem;
    font-weight:bold;
`;

const TextContainer=styled.div`
    margin-top:20px;
`;

const TextInput = styled.input.attrs({type:'text'})`
    width:100%;
    padding: 10px 15px;
    font-size:1.2rem;
    border:none;
    border-bottom:3px solid #e6b3cc;
    border-top-left-radius:5px;
    border-top-right-radius:5px;
    margin: 15px 0  40px 0; 
    
    transition: 0.2s background-color ease-in-out;
    background-color:inherit;
    color:inherit;

    &:focus{
        background-color:rgba(230, 179, 204,0.3);
    }
`;

const ButtonWrapper = styled.div`
    text-align:center;
`;

const BadgeContainer = styled.div`
    display:flex;
    flex-direction:row;
    justify-content:space-evenly;
    align-items:center;

    margin-top:20px;
`;

export default TextEditor;