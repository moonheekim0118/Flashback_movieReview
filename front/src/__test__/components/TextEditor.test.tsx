import 'jsdom-global/register';
import React from 'react';
import { ReviewProp } from './mock.data';
import { useSelector } from 'react-redux';
import { mount } from 'enzyme';
import TextEditor from '../../components/Review/TextEditor';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('<TextEditor/>', () => {
  let container;
  beforeEach(() => {
    const state = {
      myReviews: [ReviewProp, ReviewProp],
      addMyReviewDone: false,
      addMyReviewError: null,
      updateMyReviewDone: false,
      updateMyReviewError: null,
    };
    useSelector.mockImplementation(() => state);
    container = mount(<TextEditor Review={ReviewProp} ButtonType="create" />);
  });

  it('should match correctly', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
