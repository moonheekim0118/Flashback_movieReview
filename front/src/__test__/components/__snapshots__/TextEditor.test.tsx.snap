// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor/> should match correctly 1`] = `"<div class=\\"TextEditor__Container-itwvoa-0 kVrNHw\\"><div class=\\"MovieCard__Container-sc-1v9tm0k-0 hPqAwq\\"><img src=\\"test\\" alt=\\"movie poster\\" class=\\"MovieCard__MoviePoster-sc-1v9tm0k-1 hGSCZO\\"><div class=\\"MovieCard__MovieDescription-sc-1v9tm0k-2 jyXyWd\\"><p class=\\"MovieCard__MovieTitle-sc-1v9tm0k-3 irUCoO\\">test</p><p class=\\"MovieCard__MovieInfo-sc-1v9tm0k-4 bWImsu\\">test 감독</p><p class=\\"MovieCard__MovieInfo-sc-1v9tm0k-4 bWImsu\\">test 제작</p></div></div><div class=\\"TextEditor__BadgeContainer-itwvoa-5 jbrASI\\"><div class=\\"Badge__Container-sc-1qprtwz-0 kPfNfx\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"thumbs-up\\" class=\\"svg-inline--fa fa-thumbs-up fa-w-16 Icons__StyledIcon-khcp93-0 bOobqK GOOD\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" iconsize=\\"24\\" color=\\"purple\\"><path fill=\\"currentColor\\" d=\\"M104 224H24c-13.255 0-24 10.745-24 24v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V248c0-13.255-10.745-24-24-24zM64 472c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zM384 81.452c0 42.416-25.97 66.208-33.277 94.548h101.723c33.397 0 59.397 27.746 59.553 58.098.084 17.938-7.546 37.249-19.439 49.197l-.11.11c9.836 23.337 8.237 56.037-9.308 79.469 8.681 25.895-.069 57.704-16.382 74.757 4.298 17.598 2.244 32.575-6.148 44.632C440.202 511.587 389.616 512 346.839 512l-2.845-.001c-48.287-.017-87.806-17.598-119.56-31.725-15.957-7.099-36.821-15.887-52.651-16.178-6.54-.12-11.783-5.457-11.783-11.998v-213.77c0-3.2 1.282-6.271 3.558-8.521 39.614-39.144 56.648-80.587 89.117-113.111 14.804-14.832 20.188-37.236 25.393-58.902C282.515 39.293 291.817 0 312 0c24 0 72 8 72 81.452z\\"></path></svg>좋아요!</div><div class=\\"Badge__Container-sc-1qprtwz-0 INHOJ\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"meh-rolling-eyes\\" class=\\"svg-inline--fa fa-meh-rolling-eyes fa-w-16 Icons__StyledIcon-khcp93-0 bOobqK SOSO\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 496 512\\" iconsize=\\"24\\" color=\\"purple\\"><path fill=\\"currentColor\\" d=\\"M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zM88 224c0-24.3 13.7-45.2 33.6-56-.7 2.6-1.6 5.2-1.6 8 0 17.7 14.3 32 32 32s32-14.3 32-32c0-2.8-.9-5.4-1.6-8 19.9 10.8 33.6 31.7 33.6 56 0 35.3-28.7 64-64 64s-64-28.7-64-64zm224 176H184c-21.2 0-21.2-32 0-32h128c21.2 0 21.2 32 0 32zm32-112c-35.3 0-64-28.7-64-64 0-24.3 13.7-45.2 33.6-56-.7 2.6-1.6 5.2-1.6 8 0 17.7 14.3 32 32 32s32-14.3 32-32c0-2.8-.9-5.4-1.6-8 19.9 10.8 33.6 31.7 33.6 56 0 35.3-28.7 64-64 64z\\"></path></svg>보통이에요</div><div class=\\"Badge__Container-sc-1qprtwz-0 INHOJ\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"thumbs-down\\" class=\\"svg-inline--fa fa-thumbs-down fa-w-16 Icons__StyledIcon-khcp93-0 bOobqK BAD\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" iconsize=\\"24\\" color=\\"purple\\"><path fill=\\"currentColor\\" d=\\"M0 56v240c0 13.255 10.745 24 24 24h80c13.255 0 24-10.745 24-24V56c0-13.255-10.745-24-24-24H24C10.745 32 0 42.745 0 56zm40 200c0-13.255 10.745-24 24-24s24 10.745 24 24-10.745 24-24 24-24-10.745-24-24zm272 256c-20.183 0-29.485-39.293-33.931-57.795-5.206-21.666-10.589-44.07-25.393-58.902-32.469-32.524-49.503-73.967-89.117-113.111a11.98 11.98 0 0 1-3.558-8.521V59.901c0-6.541 5.243-11.878 11.783-11.998 15.831-.29 36.694-9.079 52.651-16.178C256.189 17.598 295.709.017 343.995 0h2.844c42.777 0 93.363.413 113.774 29.737 8.392 12.057 10.446 27.034 6.148 44.632 16.312 17.053 25.063 48.863 16.382 74.757 17.544 23.432 19.143 56.132 9.308 79.469l.11.11c11.893 11.949 19.523 31.259 19.439 49.197-.156 30.352-26.157 58.098-59.553 58.098H350.723C358.03 364.34 384 388.132 384 430.548 384 504 336 512 312 512z\\"></path></svg>별로에요</div></div><div class=\\"TextEditor__TextContainer-itwvoa-2 eSLYko\\"><label for=\\"shortComment\\" class=\\"TextEditor__Question-itwvoa-1 kPQRWs\\">한 줄평</label><input name=\\"shortComment\\" type=\\"text\\" class=\\"TextEditor__TextInput-itwvoa-3 bzQCDH\\" value=\\"test\\"><label for=\\"character\\" class=\\"TextEditor__Question-itwvoa-1 kPQRWs\\">기억에 남는 인물과 이유를 적어주세요</label><input name=\\"character\\" type=\\"text\\" class=\\"TextEditor__TextInput-itwvoa-3 bzQCDH\\" value=\\"test\\"><label for=\\"line\\" class=\\"TextEditor__Question-itwvoa-1 kPQRWs\\">기억에 남는 대사와 이유를 적어주세요</label><input name=\\"line\\" type=\\"text\\" class=\\"TextEditor__TextInput-itwvoa-3 bzQCDH\\" value=\\"test\\"><label for=\\"scene\\" class=\\"TextEditor__Question-itwvoa-1 kPQRWs\\">기억에 남는 장면과 이유를 적어주세요</label><input name=\\"scene\\" type=\\"text\\" class=\\"TextEditor__TextInput-itwvoa-3 bzQCDH\\" value=\\"test\\"><label for=\\"freeComment\\" class=\\"TextEditor__Question-itwvoa-1 kPQRWs\\">영화에 대한 자유로운 메모</label><input name=\\"freeComment\\" type=\\"text\\" class=\\"TextEditor__TextInput-itwvoa-3 bzQCDH\\" value=\\"test\\"></div><div class=\\"TextEditor__ButtonWrapper-itwvoa-4 rdQiR\\"><button fill=\\"false\\" color=\\"purple\\" class=\\"Buttons__Button-dz1ezt-0 dofmxo\\">저장하기</button></div></div>"`;
